java/src/org/openqa/selenium/grid/sessionmap/local/LocalSessionMap.java

```diff
@@ -48,43 +48,34 @@ public class LocalSessionMap extends SessionMap {
   private static final Logger LOG = Logger.getLogger(LocalSessionMap.class.getName());
 
   private final EventBus bus;
-  private final Map<SessionId, Session> knownSessions = new ConcurrentHashMap<>();
-  private final ReadWriteLock lock = new ReentrantReadWriteLock(/* be fair */ true);
+  private final ConcurrentMap<SessionId, Session> knownSessions = new ConcurrentHashMap<>();
 
   public LocalSessionMap(Tracer tracer, EventBus bus) {
     super(tracer);
 
     this.bus = Require.nonNull("Event bus", bus);
 
-    bus.addListener(
-        SessionClosedEvent.listener(
-            id -> {
-              try (Span span = tracer.getCurrentContext().createSpan("local_sessionmap.remove")) {
-                AttributeMap attributeMap = tracer.createAttributeMap();
-                attributeMap.put(AttributeKey.LOGGER_CLASS.getKey(), getClass().getName());
-                SESSION_ID.accept(span, id);
-                SESSION_ID_EVENT.accept(attributeMap, id);
-                knownSessions.remove(id);
-                String sessionDeletedMessage = "Deleted session from local Session Map";
-                span.addEvent(sessionDeletedMessage, attributeMap);
-                LOG.info(String.format("%s, Id: %s", sessionDeletedMessage, id));
-              }
-            }));
+    bus.addListener(SessionClosedEvent.listener(this::remove));
 
     bus.addListener(
         NodeRemovedEvent.listener(
             nodeStatus ->
                 nodeStatus.getSlots().stream()
                     .filter(slot -> slot.getSession() != null)
                     .map(slot -> slot.getSession().getId())
-                    .forEach(knownSessions::remove)));
+                    .forEach(this::remove)));
 
     bus.addListener(
         NodeRestartedEvent.listener(
-            nodeStatus ->
-                knownSessions
-                    .values()
-                    .removeIf(value -> value.getUri().equals(nodeStatus.getExternalUri()))));
+            nodeStatus -> {
+              List<SessionId> toRemove =
+                  knownSessions.entrySet().stream()
```

Comment:
`knownSessions` may be modified during iteration over the `Stream`. That's why you need to have a lock around this to ensure correctness (while the underlying type can handle concurrent insertions and deletions, the iterators will throw exceptions)

Reply:
Thanks for the hint, i will restore the lock.

Reply:
@shs96c i had a look at the [javadoc](https://docs.oracle.com/en%2Fjava%2Fjavase%2F11%2Fdocs%2Fapi%2F%2F/java.base/java/util/concurrent/ConcurrentHashMap.html), it should be safe: 
``Similarly, Iterators, Spliterators and Enumerations return elements reflecting the state of the hash table at some point at or since the creation of the iterator/enumeration. They do not throw ConcurrentModificationException. However, iterators are designed to be used by only one thread at a time.``